src/backend/distributed/planner/multi_planner.c

```diff
@@ -658,11 +659,29 @@ multi_join_restriction_hook(PlannerInfo *root,
 							JoinType jointype,
 							JoinPathExtraData *extra)
 {
+	JoinRestriction *joinRestriction = NULL;
 	JoinRestrictionContext *joinContext = NULL;
-	JoinRestriction *joinRestriction = palloc0(sizeof(JoinRestriction));
 	List *restrictInfoList = NIL;
+	MemoryContext restrictionsMemoryContext = CurrentRestrictionsMemoryContext();
+	MemoryContext oldMemoryContext = CurrentMemoryContext;
 
-	restrictInfoList = extra->restrictlist;
+	/*
+	 * Before reaching this point, some code blocks (for example GEQO optimizations)
+	 * may switch to their own memory context. If we allocate the memory here and the
+	 * lappend calls below in their context, allocated memory will be freed with that
+	 * code block's memory context. We need longer lifespan for these variables, so we
+	 * allocate them in the CurrentRestrictionsMemoryContext(), which is set to
+	 * CurrentMemoryContext at CreateAndPushPlannerRestrictionContext.
+	 */
+	oldMemoryContext = MemoryContextSwitchTo(restrictionsMemoryContext);
```

Comment:
I'd actually shorten that to "Use a memory context that's guaranteed to live long enough, could be called in a more shorted lived one (e.g. with GEQO).".